[ENH] Add tag to cli for "only" content

jdanke opened the pull request:
Simple addition to the cli to surface the sphinx "tag" command line option to build conditional content. Works like using `sphinx-build -t TAG <sourcedir>` with sphinx>=0.6 (see [sphinx documentation of the `open` directive](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-only).

This addresses issue #1290 (mistakenly refers to "open" directive). Possibly also #1425.

Multiple option allows for boolean TAG expression. For example

```jupyter-book build <sourcedir> --tag cowboy --tag cowgirl```

Is needed to build the following:
````
    ```{only} cowboy and cowgirl
     foo
    ```
````

choldgraf commented:
Thanks for this contribution - I agree that this would be valuable and in general this looks good with the tests+docs!

My one question is whether we want to special-case sphinx configuration like this, or instead make it possible for somebody to provide their own custom configuration to the `sphinx_build` command for power users who understand Sphinx well enough to make use of it. That might have potentially resolved your problem without needing to change Jupyter Book itself. On the other hand, maybe the *tag* approach is generally useful enough that others would benefit from making this explicitly supported / documented. What do others think?

jdanke commented:
Thanks for the comment. Forgive me for the following length. Being specific to get good guidance without trial & error PRs.

TL/DR: what if `config.yaml` handles the specification, and I post-process the Sphinx object in `sphinx_build` to manipulate tags, similar to how `latex_documents` are configured?

Totally agree with the question of whether adding `tag` to the cli fits the design. The advantage is minimal change to the Jupyter Book code. That may not be a sufficient reason. I get it. Also, it's a partial solution that only adds tags.

If Jupyter Book thinks it's useful to surface "tag manipulation" to the user, another approach is to work through `_config.yml` to keep it easy.

The context is that the tags are set, added, and removed during [initialization](https://github.com/sphinx-doc/sphinx/blob/799385f5558a888d1a143bf703d06b66d6717fe4/sphinx/application.py#L201) of a Sphinx object. Sphinx achieves this by executing `conf.py` to call the Tags.add() and Tags.remove() [methods](https://github.com/sphinx-doc/sphinx/blob/799385f5558a888d1a143bf703d06b66d6717fe4/sphinx/util/tags.py#L57). That's why `conf.py` contains lines like `tags.add('cowboy')` and tag entries in the [`confoverrides`](https://github.com/executablebooks/jupyter-book/blob/04fdc37fc246d44a53c6b231c3d8566d2b0f305e/jupyter_book/sphinx.py#L120) dict passed to `build_sphinx` result in warnings when building (e.g., `WARNING: unknown config value 'tags_add' in override, ignoring`).

Start with something easy: converting JB's `_config.yml` to sphinx's `conf.py`.

A user could add/remove tags in `_config.yml` like
```sphinx:
  config                    :   # key-value pairs to directly over-ride the Sphinx configuration
    html_theme              : sphinx_book_theme
    tags_add                : cowboy
    tags_remove             : cowgirl
```
Those keys are in the dict returned by `get_final_config`. To create a sphinx `conf.py`, [`cli.main.sphinx()#L476`](https://github.com/executablebooks/jupyter-book/blob/04fdc37fc246d44a53c6b231c3d8566d2b0f305e/jupyter_book/cli/main.py#L476) is modified like so. Easy.
```
    # sphinx syntax is different for tags
    tags_add = sphinx_config.pop("tags_add", None)
    tags_remove = sphinx_config.pop("tags_remove", None)

    for key in sorted(sphinx_config):
        lines.append(f"{key} = {sphinx_config[key]!r}")

    if tags_add:
        if not isinstance(tags_add, list):
            tags_add = [tags_add]
        
        for tag in tags_add:
            lines.append(f"tags.add('{tag}')")

    if tags_remove:
        if not isinstance(tags_remove, list):
            tags_remove = [tags_remove]
        
        for tag in tags_remove:
            lines.append(f"tags.remove('{tag}')")  
```

With that `conf.py` and its location `confdir` building with sphinx in [sphinx.build_sphinx()](https://github.com/executablebooks/jupyter-book/blob/04fdc37fc246d44a53c6b231c3d8566d2b0f305e/jupyter_book/sphinx.py#L113) (below) would handle adding and removing tags, but is prevented because of `patch_docutils`. I don't know why that is there, so it's probably not wise to tinker with it.
```
        with patch_docutils(confdir), docutils_namespace():
            app = Sphinx(
                srcdir=sourcedir,
                confdir=confdir,
                outdir=outputdir,
                doctreedir=doctreedir,
                buildername=builder,
                confoverrides=sphinx_config,
```

I suggest doing a similar trick as above: popping the tag-related tags from `sphinx_config` prior to instantiation to avoid the warning, then adding code to add/remove tags to `app.tags.tags` before calling `app.build(force_all, filenames)`, just like how `default_latex_documents` is handled.

jdanke commented:
Okay, I'll submit a different PR that work with the configuration yaml instead of the command line. 

Do you close this or do I?

choldgraf commented:
@jdanke - sounds good, I'll close this one and take a look at https://github.com/executablebooks/jupyter-book/pull/1630

I must admit it is a bit tough to wrap my head around all of this but I'll give it a shot haha, I've never used this tags functionality in Sphinx before!
